mlflow2seldon/api.py

```diff
@@ -0,0 +1,291 @@
+import os
+import re
+import signal
+import subprocess
+import sys
+import tempfile
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Dict
+
+import yaml
+from mlflow.exceptions import MlflowException
+from mlflow.tracking.client import MlflowClient
+from neuro_sdk import Client, Factory
+from yarl import URL
+
+
+DELAY = 2
+CLUSTER_FROM_JOB_URL_PATTERN = "jobs.(.*).org.neu.ro"
+
+
+@dataclass
+class _DeployedModel:
+    image: str
+    model_name: str
+    model_storage_uri: URL
+    model_stage: str
+    model_version: str
+    source_run_id: str
+    deployment_namespace: str
+    need_redeploy: bool = False
+
+    def is_same_version(self, other: _DeployedModel) -> bool:
+        all_is_true = all(
+            (
+                self.name == other.name,
+                self.model_version == other.model_version,
+            )
+        )
+        return all_is_true
+
+    @property
+    def name(self) -> str:
+        return f"{self.model_name}-{self.model_stage}".lower()
+
+
+async def poll_mlflow(env: Dict[str, str]) -> None:
+
+    # Settings of the source cluster, where MLflow is deployed:
+    mlflow_storage_root = env["M2S_MLFLOW_STORAGE_ROOT"]
+    mlflow_host = env["M2S_MLFLOW_HOST"]
+    default_image = env["M2S_SELDON_NEURO_DEF_IMAGE"]
+    deploy_image_tag = env["M2S_MLFLOW_DEPLOY_IMG_TAG"]
+    registry_secret_name = env["M2S_NEURO_REGISTRY_SECRET"]
+    seldon_deployment_ns = env["M2S_SELDON_DEPLOYMENT_NS"]
+
+    client_factory = Factory()
+    # Assumption: MLFlow is running in a platform job
+    cluster = re.findall(CLUSTER_FROM_JOB_URL_PATTERN, mlflow_host)[0]
+    neuro_client = await client_factory.get()
+    await neuro_client.config.switch_cluster(cluster)
+    mlflow_client = MlflowClient(tracking_uri=mlflow_host)
+    # Assumption: Seldon deployment image is on a same cluster, where MLflow is running
+    default_image_ref = neuro_client.parse.remote_image(default_image).as_docker_url()
+
+    seldon_models: Dict[str, _DeployedModel] = dict()
+    try:
+        while True:
+            mlflow_models: Dict[str, _DeployedModel] = dict()
+            logging.info(f"Polling {mlflow_host}")
+            try:
+                # iterate over all registered models ("Models" tab in MLflow WebUI)
+                for model in mlflow_client.search_registered_models():
+                    deploy_image = model.tags.get(deploy_image_tag)
+                    if deploy_image:
+                        deploy_image_ref = neuro_client.parse.remote_image(
+                            deploy_image
+                        ).as_docker_url()
+                    else:
+                        deploy_image_ref = default_image_ref
+                    for model_version in model.latest_versions:
+                        if model_version.current_stage not in ("Staging", "Production"):
+                            # we deploy only Staging and Production models
+                            continue
```

Comment:
here: not `continue` but `kubectl delete --ignore-not-found`: see https://github.com/dmangonakis/mlops-lab-example-yelp/blob/main/model_operator/app/main.py#L33-L44

Reply:
the idea is to implement a full synchronisation: MLflow -> Seldon. Now we only `kubectl apply` the registered deployments, but don't "sync"

Reply:
you mean to delete those models, that first were in `staging` / `production` state, but later got `archived` ? 
those ones are deleted [here](https://github.com/neuro-inc/mlops-k8s-mlflow2seldon/blob/ys/refactor/mlflow2seldon/api.py#L130):
var `seldon_models` holds the state of previous sync, where currently outdated model existed, but not now

Reply:
hmm okay this deletes the models that were obsolete since the hellgate has started, but not the models that were running before. It's ok for the demo

Reply:
yeah, I completely agree with you, I keep this thing tracked here: #9